<b>Import the required packages</b>

In [13]:
import os
import glob
import nltk
from sklearn import svm
from sklearn.feature_extraction.text import CountVectorizer, TfidfTransformer
from sklearn.metrics import accuracy_score

<b> We first write a function to preprocess the data which does below:</b> <br>
1) We first define some unwanted characters. <br>
2) Then we tokenize the data.<br>
3) Then we remove the stop words and unwanted characters that we defined and stem the final output.<br>
4) then we join the tokens together to form a string and return it.<br>


In [14]:


unwanted = ['\n','\t','\r','\'','"',':',';','.','/','!','#','$','%','{','}','(',')','?','&',']','[','-','_']


def pre(text1):
 lis1=nltk.word_tokenize(text1)
 lis1=[nltk.stem.PorterStemmer().stem(word.lower()) for word in lis1 if word not in nltk.corpus.stopwords.words("English") and word not in unwanted]
 text=" ".join(lis1)
 return text

1) Then using the preprocessing function above we append preprocessed files to respective corpus: file_ham and file_spam <br>
2) We also create a list with label for each item in corpus in the name of labels.

In [15]:
file_ham=[]
for filename in glob.glob(os.path.join("./MailDataset/ham", '*.txt')):
 file=open(filename,'r')
 file_ham.append(pre(file.read()))


file_spam=[]
for filename in glob.glob(os.path.join("./MailDataset/spam", '*.txt')):
 file=open(filename,'r')
 file_spam.append(pre(file.read()))


labels=[]

for i in range(0,200):
    if i<100:
        labels.append("ham")
    else:
        labels.append("spam")

1) Using the above dataset we <b>train</b> the model with <b>60% data</b><br>
2) We first convert the data into tfidf vectors.<br>
3) Then we define the svm model and fit training data to it with different values of hyperparameters and finally use  gamma=.001, C=4 after multiple performance tests.<br>  

In [16]:
corpus=file_ham[40:100]+file_spam[:60]
print(len(corpus))
count_vect=CountVectorizer()
tfidf_transformer=TfidfTransformer()
counts=count_vect.fit_transform(corpus)
Z=tfidf_transformer.fit_transform(counts)

model=svm.SVC(gamma=.001, C=4)
model.fit(Z,labels[:60]+labels[140:])

120


SVC(C=4, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape='ovr', degree=3, gamma=0.001, kernel='rbf',
  max_iter=-1, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False)

<br>Then we use 40% data to test the model and use accuracy as the metric to measure the performance.

In [17]:

counter=count_vect.transform(file_ham[:40]+file_spam[60:100])
predict=tfidf_transformer.transform(counter)


model.predict(predict)
predicted=model.predict(predict)
print(accuracy_score(predicted,(labels[60:100]+labels[100:140])))

0.8625


As can be seen above the accuracy is 86.25% which is good considering the data that we have.